Integrate Kotlin runtime

KvanTTT opened the pull request:
I've merged Strumenta's Kotlin Runtime repository (git history is preserved) and integrated it in out ANTLR runtime test infrastructure ([`runtime-testsuite`](https://github.com/antlr/antlr5/tree/dev/runtime-testsuite)). Fixes #1 

To make it possible to run tests, I had to fix `Kotlin.stg` template, fix some runtime files, write `Kotlin.test.stg` and other wrapping code. Also, I had to remove `expect/actual` functions because they don't work reliable with Maven, and `expect/actual` within one module are prohibited since Kotlin 2.0.

Generally, everything is not so bad as I thought before, and currently we have 35 failed tests out of 356. It's worth mention, `kotlinc` cmd tools should be configured.

Remaining problems in tests:

* Incorrect declaring of some labels, for instance [ParserExec/Labels](https://github.com/antlr/antlr5/blob/dev/runtime-testsuite/resources/org/antlr/v5/test/runtime/descriptors/ParserExec/Labels.txt) and other:

```
error: extension property cannot be initialized because it has no backing field.
        public var Tokens.INT: Token? = null
```

Subsystem to fix: code generation.

* Clash because of same JVM signatures, for instance [ParserErrors/SingleSetInsertionConsumption](https://github.com/antlr/antlr5/blob/dev/runtime-testsuite/resources/org/antlr/v5/test/runtime/descriptors/ParserErrors/SingleSetInsertionConsumption.txt):

```
error: platform declaration clash: The following declarations have the same JVM signature (getExpr()LTParser$ExprContext;):
    fun `<get-expr>`(): TParser.ExprContext? defined in TParser.SContext
    fun getExpr(): TParser.ExprContext defined in TParser.SContext
        public var expr: ExprContext? = null
        ^
```

Subsystem to fix: code generation.

* Escaping of reserved words ([exerExec/ReservedWordsEscaping](https://github.com/antlr/antlr5/blob/dev/runtime-testsuite/resources/org/antlr/v5/test/runtime/descriptors/LexerExec/ReservedWordsEscaping.txt), [ParserExec/ReservedWordsEscaping](https://github.com/antlr/antlr5/blob/dev/runtime-testsuite/resources/org/antlr/v5/test/runtime/descriptors/ParserExec/ReservedWordsEscaping.txt))

Subsystem to fix: code generation (probably only `Kotlin.stg`).

* Unicode issues ([descriptors/Sets](https://github.com/antlr/antlr5/tree/dev/runtime-testsuite/resources/org/antlr/v5/test/runtime/descriptors/Sets)). For instance, emojies are not correctly treated.

Subsystem to fix: Kotlin runtime.

KvanTTT commented:
@ericvergnaud is it possible to set up GitHub Actions for this repository (probably some option in GitHub UI). I don't see Actions in this repository and CI is not triggered despite the fact `.github/workflows/hosted.yml` file exists in the repository (as it exists in antlr4).

I want to make sure Kotlin GitHub Action is configured correctly and Kotlin runtime tests work.

ericvergnaud commented:
I've enabled actions

ericvergnaud commented:
Don't forget to -s your commits

lppedd commented:
Very nice, thanks!

Happy to see we don't have that many failing tests too.

KvanTTT commented:
> Don't forget to -s your commits

I have signature (if I understand `-s` correctly) on all my commits. But I merged a lot of other commits without signature from external Kotlin repository. Probably it affects DCO bot as well.

KvanTTT commented:
Interesting, GitHub CI is not activated for this PR anyway.

KvanTTT commented:
@ericvergnaud it could be caused by [merge conflicts](https://github.com/orgs/community/discussions/26304). Sorry, but may I drop all your commits in ANTLR5 from `dev` since ebb511a04a60ae5a605aba65471c07dd854e9303 (your latest commit in antlr4)? I started my Kotlin integration branch on this commit and we have definitely working build on this commit unlike the current `dev`.

I suggest making renaming and versions upgrading a bit later, I'm not sure it's very improtant for now. But I'll keep your already introduced commits separately in another branch and nothing will be loosed (probably you also have this branch unremoved). Also, next time I'm going to squash them to get rid of non-informative commits.

ericvergnaud commented:
Sure, whatever works is fine with me, provided we don't lose any changes

ftomassetti commented:
Thank you @KvanTTT : that is amazing news. I am running the tests locally and try to see how I can help get some of these tests fixed

KvanTTT commented:
Now Kotlin tests work on CI at least on Ubuntu: https://github.com/antlr/antlr5/actions/runs/7643931630/job/20827098759?pr=6#step:24:4436 (it shows 36 failed tests instead of 35 as I wrote before, but locally I tested on K2 and it doesn't fail on a big grammar unlike the current K1).

KvanTTT commented:
The bad news: Kotlin tests are very slow, it's especially actual for remote servers. But I have an idea how performance can be significantly improved. I'll try it later (it looks like the full runtime is being compiled on every test, it's very wasteful, much better to compile the runime only once and use result class files in tests as we do for other runtimes).

ftomassetti commented:
I have prepared a PR on this PR (#8).

It brings down the tests failing from 36 to 3.

They are:
* LexerExec.AtnStatesSizeMoreThan65535:  we have an OutOfMemoryError here, but if I understood correctly what @KvanTTT wrote, it will disappear when using K2
* Sets.UnicodeEscapedSMPRangeSetMismatch: this one is tricky, as if we compile to WASM, I guess we cannot use Java specific classes for dealing with Unicode, so I am not sure how to solve this one
* Sets.UnicodeEscapedSMPRangeSet: same as the one before

KvanTTT commented:
> It brings down the tests failing from 36 to 3.

Great news!

> LexerExec.AtnStatesSizeMoreThan65535: we have an OutOfMemoryError here, but if I understood correctly what @KvanTTT wrote, it will disappear when using K2

Yes, I meant exactly that. This test is green if switch to K2.

> Sets.UnicodeEscapedSMPRangeSetMismatch: this one is tricky, as if we compile to WASM, I guess we cannot use Java specific classes for dealing with Unicode, so I am not sure how to solve this one

I'll take a look what we can do with Unicode text once again. If I fails, I think Unicode tests can be ignored.

ericvergnaud commented on tool/src/org/antlr/v4/codegen/target/KotlinTarget.java:
```diff
@@ -84,7 +84,7 @@ public Map<Character, String> getTargetCharValueEscape() {
 		// All tokens are namespaced inside a Tokens object.
 		// Here we simply force the qualification
-		return "Tokens." + super.getTokenTypeAsTargetLabel(g, ttype);
+		return super.getTokenTypeAsTargetLabel(g, ttype);
```
Maybe remove this method altogether ?

KvanTTT replied:
Sure, I've already removed it in my local branch. Currently I'm figuring out if I can fix two remaining failing tests related to Unicode stuff.

KvanTTT commented:
So, I've fixed all remaining tests, now build status is completely green:

* Switching to K2 fixed a bug with big generated files (`AtnStatesSizeMoreThan65535`)
* Tests with Unicode stuff were fixed by fixing [`StringCharStream.kt`](https://github.com/antlr/antlr5/commit/ab0c3f5a8dc8f11ff9af28391879febdccfc59d5) (turned out it's not a jvm-specific issue but a bug in `getText` function)

Also, I've fixed testsuite running on Windows CI.

Finally, I've integrated Strumenta's [Kotlin runtime specific tests](https://github.com/antlr/antlr5/commit/abb36a8267c5456f66d4a9ad13c6b995f632c3e1). I'm not sure if all of them are valuable considering the fact now we have much better coverage, but we can remove them later if it's needed.

@ericvergnaud I think it's ready for review. If everything is ok, just **Approve** the PR and I'll merge it by myself (it's not a regular PR since it contains a merge from unrelated histories and it's better to merge it manually).

ericvergnaud commented:
Wow! Amazing! I'll look into that tomorrow morning 😃 !

KvanTTT commented on .github/workflows/hosted.yml:
```diff
@@ -166,16 +166,17 @@ jobs:
-          typescript,
-          php,
-          python3,
+          # cpp,
+          # csharp,
+          # dart,
+          # go,
+          # java,
+          # javascript,
+          # typescript,
+          # php,
+          # python3,
```
I'll uncomment it before the final merge.

I commented it out for more faster remote building when I was trying to fix Windows CI.

KvanTTT commented:
I leaved untouched unused `kt` files in [antlr-kotlin-runtime/src](https://github.com/antlr/antlr5/tree/integrate-kotlin-runtime/antlr-kotlin-runtime/src):

* jsAndWasmSharedMain
* jsMain
* nativeMain
* wasmJsMain
* wasmWasiMain

It we are going to implement KMP features later (with Native, Wasm targets), these files can be used and the git history will be preserved as currently it's preserved for Kotlin runtime files (git is able to track files moving from directory to directory).

ericvergnaud commented on LICENSE-Apache:
```diff
@@ -0,0 +1,203 @@
+
```
All Kotlin runtime contributors have agreed to also license their code under BSD-3, which is the overarching license in this repo, so this file should be removed.

ericvergnaud commented on LICENSE-BSD:
```diff
@@ -0,0 +1,11 @@
+Copyright 2024 ANTLR Kotlin contributors
```
The license is provided at the repo level, see https://github.com/antlr/antlr5/blob/dev/LICENSE.txt so this file should be removed

ericvergnaud commented on runtime-testsuite/test/org/antlr/v4/test/runtime/PredictionMode.java:
```diff
@@ -0,0 +1,7 @@
+package org.antlr.v4.test.runtime;
```
Shouldn't we be retrieving this enum from the runtime ?

KvanTTT replied:
No, it should not. Java and Kotlin runtimes have their own `PredictionMode` declarations. It's not correctly to use Java runtime enum in Kotlin runtime tests.

ericvergnaud replied:
Not sure I follow, if it's not correct to use Java runtime enum in Kotlin runtime tests, how is it correct to use a generic test-suite enum in these Kotlin runtime tests ? (I'm genuinely trying to understand)

KvanTTT replied:
We use [RuntimeTestDescriptor](https://github.com/antlr/antlr5/blob/dev/runtime-testsuite/test/org/antlr/v4/test/runtime/RuntimeTestDescriptor.java) as a descriptor that stores info about `.txt` and custom tests. It actually describes common data related to all runtimes, but not only to Java-runtime. If we decide to completely drop Java-runtime, this enum will be unavailable and we have to drop its using anyway from `runtime-testsuite` module. I removed it within this PR to minimize future effort and to clarify the code, see the commit https://github.com/KvanTTT/antlr4/pull/10/commits/e939f14092c786eaf48d9e27b5e8eff704a1383a Currently Kotlin test helpers don't depend on any Java-runtime entity.

ericvergnaud replied:
Makes sense

KvanTTT commented:
Merged manually (force-pushed to dev)! Now we can use standard and safe GitHub merging methods like `Rebase and merge` or `Squash and merge`.

ftomassetti commented:
Great to see this one merged!
